app/controllers/api/v1/products_controller.rb

```diff
@@ -0,0 +1,49 @@
+module Api
+  module V1
+    class ProductsController < ApplicationController
+      before_action :set_product, except: %i[create index]
+      def index
+        render json: { data: ActiveModel::SerializableResource.new(Product.all, each_serializer: ProductSerializer) }
```

Comment:
- What if we have `10_000_000` products?
- It's rather `ActiveModelSerializers::SerializableResource` if anything. It should work out of the box with Rails:
```suggestion
        render json: { data: Product.all, each_serializer: ProductSerializer }
```
https://github.com/rails-api/active_model_serializers/blob/0-10-stable/docs/general/getting_started.md#rails-integration